doc/source/development/window.rst

```diff
@@ -0,0 +1,187 @@
+
+
+An aggregation is an interface that implements methods to update its internal
+state and a :meth:`finalize` method to return the current value of the
+aggregation based on the internal state.
+
+An example implementation of a :class:`Sum` kernel would look similar to
+
+.. code-block:: python
+
+   from typing import Generic, TypeVar
+
+   Input1 = TypeVar("Input1")
+   Input2 = TypeVar("Input2")
+   Output = TypeVar("Output")
+
+   class UnaryAggregation(Generic[Input1, Output]):
+       @abc.abstractmethod
+       def step(self, value: Optional[Input1]) -> None:
+           ...
+
+   class Sum(UnaryAggregation[Input1, Output]):
+       def __init__(self):
+           self.count: int = 0
+           self.total: Output = 0
+
+       def step(self, value: Optional[Input1]) -> None:
+           if value is not None:
+               self.count += 1
+               self.total += value
+
+       def invert(self, value: Optional[Input1]) -> None:
+           """Used only in subtractable aggregator."""
+           if value is not None:
+               self.count -= 1
+               self.total -= value
+
+       def finalize(self) -> Optional[Output]:
+           if not self.count:
+               return None
+           return self.total
+
+       def combine(self, other: Sum) -> None:
+           """Used only in segment tree aggregator."""
+           self.total += other.total
+           self.count += other.count
+
+       @classmethod
+       def make_aggregator(cls, values: ndarray) -> AggregationAlgorithm:
+           SubtractableAggregator.check_agg(cls)
+           aggregator = SubtractableAggregator(values, cls())
+           return aggregator
+
+   class BinaryAggregation(Generic[Input1, Input2, Output]):
+       @abc.abstractmethod
+       def step(
+           self, value1: Optional[Input1], value2: Optional[Input2]
+       ) -> None:
+           ...
+
+   class Covariance(BinaryAggregation[Input1, Input2, float])
+       ...
+
+
+Customization of Window Spans
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+We propose a generic mechanism that allows power users and library authors to
+customize the computation of window boundaries.
+
+.. code-block:: python
+
+   import abc
+   from typing import Tuple
+
+   BeginEnd = Tuple[ndarray, ndarray]
+
+   class Indexer(abc.ABC):
+       @abc.abstractmethod
+       @classmethod
+       def get_window_bounds(cls, index, offset) -> BeginEnd:
+           ...
+
+   class MyBusinessDayIndexer(Indexer):
+       @classmethod
+       def get_window_bounds(cls, index, offset) -> BeginEnd:
+           ...
+
+Example Aggregation Loop Implemention
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Here is an example implementation of a loop that would perform the aggregation
+using the interfaces proposed above.
+
+.. code-block:: python
+
+   def do_agg(
+       values: ndarray,
+       index: ndarray,
+       offset,
+       indexer_class: Type[Indexer],
+       agg_class: Type[Agg],
```

Comment:
Agg -> AggregationAlgorithm

Reply:
No, this is a kernel. `AggregationAlgorithm`s are constructed by kernels in this function below.

Reply:
ok, that's why I was slightly confused by the Agg term